text/0000-deref-move.md

```diff
@@ -0,0 +1,325 @@
+T` to pass into the method, or it could call `deref_move` to get a `&mut T`
+directly. The former is chosen because it avoids moving the pointer. The same
+ambiguity arises with `&T`, but is already accounted for by the `Copy` rule
+since shared references are `Copy`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Nobody understands autoderef already and this makes it worse
+
+Or at least, it would if `Box` wasn't already complicating it in the same way.
+
+## Copy behaviour is unintuitive in generic cases
+
+If a type has nontrivial/side-effectful `deref_move`, it may be slightly
+confusing why it is not used for `Copy` types in non-generic code, but is used
+in generic code.
+
+## Breaking change to an edge case of `Box<&mut T>`
+
+Currently, a `Box<&mut T>` variable [can be
+dereferenced](https://play.rust-lang.org/?gist=5b3f046ecf29da8f00fde3bd1d2a3df4&version=nightly&mode=debug)
+in at least some cases into a `&mut T` without either requiring that the
+variable be mutable (as would be the case for any other `DerefMut` type), or
+moving out of the `Box`. This behaviour would not be maintained if its magical
+dereferencing were replaced with `DerefMove`, as `DerefMut` requires a mutable
+variable.
+
+This is an edge case, as it requires `Box<&mut T>`, which is an unlikely type,
+and in most cases it can be trivially fixed by making the variable mutable. It's
+possible there are more complex cases (such as if the `Box` itself is stored in
+another smart pointer) where this is not easily fixed, but I would be
+unsurprised if there were few, if any, instances of this in extant code.
+Moreover, the behaviour is arguably a bug because it is taking a mutable borrow
+of an immutable variable (and one can modify the above example to get an error
+message implying exactly that), though the semantics are more like the unique
+immutable borrows used by closures.
+
+As a result, I propose that this breaking change be accepted after verifying its
+effect on Cargo packages.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## Copy handling
+
+The handling of copies is the real complexity of `DerefMove`. Currently, one can
+copy out of a `Box` without consuming it, so this desirable behaviour needs to
+be preserved. The proposal is to do so by special casing in the compiler.
+However, it could also be done via, for instance, a separate method in the Deref
+trait definition:
+
+```rust
+   fn deref_copy(&self) -> Self::Target where Self::Target : Copy {
+       return self.deref();
+   }
+```
+
+This still requires special-casing in the compiler, but it is less magical. One
+could even envision a `DerefCopy` trait with similar purpose.
+
+These approaches would, however, make diverging implementations (that is,
+implementations for which the behaviour of the various deref traits is
+unintuitive because they don't actually refer to the same underlying value) more
+difficult to write, and we want to avoid that. And they would not fundamentally
```

Comment:
Do you mean easier to write?

Reply:
Yes, sorry. See my above comment that if we did this we may actually need to enforce that they match.

Reply:
@alercah 

> The behaviour of ignoring DerefMove when Target : Copy for Deref would be complicated if the Target types were different; that would possibly create scenarios where a generic smart pointer would fail for a non-Copy Target because DerefMove is ignored but Deref has a different Target.

Can you elaborate on what you mean by "fail"?  I assume all that would happen is that the Deref target and implementation would be used on autoderef, which seems no different than what would happen if the targets were the same.

> This would also complicate autoderef and in particular its relationship with overload resolution, probably unbearably so.

Wouldn't it always be unambiguous which implementation to use in any given context, in the same way that `Copy` itself is?  I may be missing something.  Or is the problem that the two types could have methods with the same name, and you couldn't tell which one to call or prioritize one over the other?  If that's your concern I definitely see why you would want the types to match.